Level 1 implementation

kbence opened the pull request:
Work in progress

kbence committed:
The first "naive" implementation

This basic implementation is missing some key features, eg.
detecting the direction of the connections. Please be mindful
of that!

wadells commented on go.mod:
```diff
@@ -0,0 +1,8 @@
+
+require (
+	github.com/spf13/cobra v1.1.3
+	k8s.io/utils v0.0.0-20210305010621-2afb4311ab10
```
This doesn't appear to be used.  `go mod tidy`?

kbence replied:
Ah, VSCode is trying to mess with me :D. It had accidentally included a k8s import. I didn't notice that it has also added it to `go.mod`. Thanks!

wadells commented on internal/connlib/tcp.go:
```diff
@@ -0,0 +1,85 @@
+
+	ip, err := strconv.ParseUint(parts[0], 16, 32)
+	if err != nil {
+		panic(err)
```
Panic is atypical chose for a go library function.  Why this over propagating the error upwards?

kbence replied:
Oh yeah, that was a quick and dirty one missed. Thank you!

wadells reviewed (approved):
Looks great!  Will look out for directionality later per your remark.

kbence committed:
Cleaned up unused dependencies

kbence committed:
Better error handling

This change passes back all the errors to the main command function
to be handled (more) gracefully.

kbence committed:
Fix panic on long sequence numbers

When /proc/net/tcp gets big due to the number of TCP sockets,
the sequence numbers start using up the space they are aligned
with, therefore the regular expression "\s+" is no longer
sufficient to parse the lines. Luckily, there's a ":" after the
number, so let's split the line there first.

I have a rule to introduce testing in a project when the first
bug appears, so I've already wrote the first tests to fix this.

kbence committed:
Detecting directionality

This change enables the program to guess directionality based on:
- ports listening on specific IPs
- ports listening on 0.0.0.0
- ports that are ephemeral on the system it's running on

The latter two are not exactly reliable unfortunately, since what's
an outgoing port on one node, could be a listening port on another.

The best solution would be to parse out all IP addresses assigned
to the current node from /proc/net/fib_trie, but that's not that
simple and I probably won't have time for that.

kbence committed:
Add some command line arguments

Some configurability to make it easier to test and since it's a
low hanging fruit.

wadells commented on test.sh:
```diff
@@ -0,0 +1,6 @@
+ROOT=$(cd "$(dirname ${BASH_SOURCE[0]:-$0})"; pwd)
+
+cd "$ROOT"
+go test $(find . -name '*_test.go' | sed 's/[^/]\+$//' | uniq)
```
What advantage does this `$(find ...)` provide over `./...`?

kbence replied:
I figured later on that `go test` can only run one package at the time, so I put it in a `for` loop, but I just expected to work if `*_test.go` files are popping up elsewhere. I was naive :smile:

wadells commented on internal/connlib/tcp.go:
```diff
@@ -0,0 +1,141 @@
+	// Note: this is totally not right, since we don't have any
+	// information about what IP addresses this specific node has.
+
+	// TODO: parse it out from /proc/net/fib_trie and use that
+	// list instead of IP.IsUnspecified()!
```
Thanks for these comments.  Even though the work is out of scope, it is nice to know how you'd approach the next step.

wadells reviewed (approved):
New set of patches also looks good to me!  Nice testing.
